tracing/tracer_server.go

```diff
@@ -0,0 +1,119 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+	"time"
+
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/cloudwego/kitex/pkg/stats"
+	"github.com/cloudwego/kitex/server"
+	"github.com/kitex-contrib/obs-opentelemetry/tracing/internal"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/metric"
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+var _ stats.Tracer = (*serverTracer)(nil)
+
+type serverTracer struct {
+	config            *config
+	histogramRecorder map[string]metric.Float64Histogram
+}
+
+func newServerOption(opts ...Option) (server.Option, *config) {
+	cfg := newConfig(opts)
+	st := &serverTracer{
+		config: cfg,
+	}
+
+	st.createMeasures()
+
+	return server.WithTracer(st), cfg
+}
+
+func (s *serverTracer) createMeasures() {
+	serverDurationMeasure, err := s.config.meter.NewFloat64Histogram(ServerDuration)
+	handleErr(err)
+
+	s.histogramRecorder = map[string]metric.Float64Histogram{
+		ServerDuration: serverDurationMeasure,
+	}
+}
+
+func (s *serverTracer) Start(ctx context.Context) context.Context {
+	tc := &internal.TraceCarrier{}
+	tc.SetTracer(s.config.tracer)
+
+	return internal.WithTraceCarrier(ctx, tc)
+}
+
+func (s *serverTracer) Finish(ctx context.Context) {
+	// trace carrier from context
+	tc := internal.TraceCarrierFromContext(ctx)
+	if tc == nil {
+		return
+	}
+
+	// rpc info
+	ri := rpcinfo.GetRPCInfo(ctx)
+	if ri.Stats().Level() == stats.LevelDisabled {
+		return
+	}
+
+	st := ri.Stats()
+	rpcStart := st.GetEvent(stats.RPCStart)
+	rpcFinish := st.GetEvent(stats.RPCFinish)
+	duration := rpcFinish.Time().Sub(rpcStart.Time())
+	elapsedTime := float64(duration) / float64(time.Millisecond)
+
+	// span
+	span := tc.Span()
+	if span == nil || !span.IsRecording() {
+		return
+	}
+
+	// span attributes
+	attrs := []attribute.KeyValue{
+		RPCSystemKitex,
+		RequestProtocolKey.String(ri.Config().TransportProtocol().String()),
+		semconv.RPCMethodKey.String(ri.To().Method()),
+		semconv.RPCServiceKey.String(ri.To().ServiceName()),
+		RPCSystemKitexRecvSize.Int64(int64(st.RecvSize())),
+		RPCSystemKitexSendSize.Int64(int64(st.SendSize())),
+	}
+
+	// The source operation dimension maybe cause high cardinality issues
+	if s.config.recordSourceOperation {
+		attrs = append(attrs, SourceOperationKey.String(ri.From().Method()))
+	}
+
+	span.SetAttributes(attrs...)
+
+	injectStatsEventsToSpan(span, st)
+
+	if st.Error() != nil {
+		attrs = append(attrs, StatusKey.String(codes.Error.String()))
+		RecordErrorSpan(span, st.Error(), s.config.withStackTrace, attrs...)
```

Comment:
I think it's redundant to add attribute when record err

Reply:
> I think it's redundant to add attribute when record err

Yes, this is indeed redundant. The metrics can get the status from the span to set the status dimensions

Reply:
The attributes passed to recordErrorSpan will be used to create an error event, and record such attributes on error event, for example `rpc.system` `protocal` and so on, is unnecessary.

Reply:
> The attributes passed to recordErrorSpan will be used to create an error event, and record such attributes on error event, for example `rpc.system` `protocal` and so on, is unnecessary.

Makes sense, event doesn't really need these attributes